File: lib/transdb.py
"""Database for the pending transactions pool."""

import os
from typing import Generator

import plyvel

from .crypto import Hash512
from .errors import TransactionNotFoundError
from .transaction import Transaction


class TransDB:
    """Database for pending transactions."""

    def __init__(self, base_dir: str) -> None:
        """Create a new TransDB object."""
        self.base_dir = base_dir
        self.transfile = os.path.join(self.base_dir, "trans.db")
        self.db = plyvel.DB(self.transfile, create_if_missing=True)
        return

    def bump_index(self) -> int:
        """Bump the next transaction index and return the existing one."""
        index = int(self.db.get(b'nextindex', b'0'))
        next_index = index + 1
        self.db.put(b'nextindex', str(next_index).encode('utf-8'))
        return index

    def get_key_for_index(self, index: int) -> bytes:
        """Get the key for the specified index."""
        return "trans_{:09d}".format(index).encode('utf-8')

    def get_key_for_txid(self, txid: Hash512) -> bytes:
        """Get the key for the specified transaction id."""
        txid_str = txid.to_string()
        return "txid_{}".format(txid_str).encode('utf-8')

    def get_key_bumped_index(self) -> bytes:
        """Bump the index and get the key."""
        return self.get_key_for_index(self.bump_index())

    def add_transaction(self, trans: Transaction) -> None:
        """Add the transaction to the pool."""
        data = trans.serialise().encode('utf-8')
        key = self.get_key_bumped_index()
        self.db.put(key, data)

        # Link transaction id to key.
        txkey = self.get_key_for_txid(trans.txid)
        self.db.put(txkey, key)
        return

    def get_transaction(self, txid: Hash512) -> Transaction:
        """Get the specified transaction."""
        txid_str = txid.to_string()
        txkey = self.get_key_for_txid(txid)
        key = self.db.get(txkey)
        if key is None:
            raise TransactionNotFoundError(
                "Unable to find transaction {}".format(txid_str))

        raw_data = self.db.get(key)
        if raw_data is None:
            raise TransactionNotFoundError(
                "Unable to find transaction {}".format(txid_str))

        return Transaction.deserialise(raw_data)

    def delete_transaction(self, txid: Hash512) -> None:
        """Delete the specified transaction from the pending list."""
        txid_str = txid.to_string()
        txkey = self.get_key_for_txid(txid)
        key = self.db.get(txkey)
        if key is None:
            raise TransactionNotFoundError(
                "Unable to find transaction {}".format(txid_str))

        with self.db.write_batch() as wb:
            wb.delete(key)
            wb.delete(txkey)
        return

    def get_transaction_iterator(self) -> Generator[Transaction, None, None]:
        """Get list of transactions."""
        for raw_data in self.db.iterator(prefix=b'trans_', include_key=False):
            tx = Transaction.deserialise(raw_data)
            yield tx
        return

File: rs/src/transdb.rs
/// Database for the pending transactions pool — direct port of lib/transdb.py.
use rusty_leveldb::DB;

use crate::crypto::Hash;
use crate::errors::{BlockChainError, Result};
use crate::transaction::Transaction;

pub struct TransDB {
    pub db: DB,
}

impl TransDB {
    pub fn new(base_dir: &str) -> Result<Self> {
        let transfile = format!("{base_dir}/trans.db");
        let opts = rusty_leveldb::Options::default();
        let db = DB::open(&transfile, opts)
            .map_err(|e| BlockChainError::DbIntegrity(format!("Failed to open TransDB: {e}")))?;
        Ok(TransDB { db })
    }

    fn bump_index(&mut self) -> u64 {
        let index: u64 = self
            .db
            .get(b"nextindex")
            .and_then(|v| String::from_utf8(v).ok())
            .and_then(|s| s.parse().ok())
            .unwrap_or(0);

        let next = index + 1;
        let _ = self.db.put(b"nextindex", next.to_string().as_bytes());
        index
    }

    fn get_key_for_index(index: u64) -> Vec<u8> {
        format!("trans_{index:09}").into_bytes()
    }

    fn get_key_for_txid(txid: &Hash) -> Vec<u8> {
        format!("txid_{}", txid.to_hex()).into_bytes()
    }

    pub fn add_transaction(&mut self, trans: &Transaction) -> Result<()> {
        let data = trans.serialise()?.into_bytes();
        let index = self.bump_index();
        let key = Self::get_key_for_index(index);

        self.db
            .put(&key, &data)
            .map_err(|e| BlockChainError::DbIntegrity(format!("DB write failed: {e}")))?;

        // Link txid -> key
        let txkey = Self::get_key_for_txid(&trans.txid);
        self.db
            .put(&txkey, &key)
            .map_err(|e| BlockChainError::DbIntegrity(format!("DB write failed: {e}")))?;
        Ok(())
    }

    pub fn get_transaction(&mut self, txid: &Hash) -> Result<Transaction> {
        let txkey = Self::get_key_for_txid(txid);
        let key = self.db.get(&txkey).ok_or_else(|| {
            BlockChainError::TransactionNotFound(format!(
                "Unable to find transaction {}",
                txid.to_hex()
            ))
        })?;

        let raw_data = self.db.get(&key).ok_or_else(|| {
            BlockChainError::TransactionNotFound(format!(
                "Unable to find transaction {}",
                txid.to_hex()
            ))
        })?;

        let json_str = String::from_utf8(raw_data).map_err(|e| {
            BlockChainError::TransactionNotFound(format!("Invalid UTF-8 transaction data: {e}"))
        })?;
        Transaction::deserialise(&json_str)
    }

    pub fn delete_transaction(&mut self, txid: &Hash) -> Result<()> {
        let txkey = Self::get_key_for_txid(txid);
        let key = self.db.get(&txkey).ok_or_else(|| {
            BlockChainError::TransactionNotFound(format!(
                "Unable to find transaction {}",
                txid.to_hex()
            ))
        })?;

        self.db
            .delete(&key)
            .map_err(|e| BlockChainError::DbIntegrity(format!("DB delete failed: {e}")))?;
        self.db
            .delete(&txkey)
            .map_err(|e| BlockChainError::DbIntegrity(format!("DB delete failed: {e}")))?;
        Ok(())
    }

    pub fn get_pending_transactions(&mut self) -> Vec<Transaction> {
        let prefix = b"trans_";
        let mut txs = Vec::new();

        use rusty_leveldb::LdbIterator;
        if let Ok(mut iter) = self.db.new_iter() {
            iter.seek(prefix);
            let mut key_buf = Vec::new();
            let mut val_buf = Vec::new();
            loop {
                key_buf.clear();
                val_buf.clear();
                if !iter.valid() {
                    break;
                }
                if iter.current(&mut key_buf, &mut val_buf) {
                    if !key_buf.starts_with(prefix) {
                        break;
                    }
                    if let Ok(s) = String::from_utf8(val_buf.clone()) {
                        if let Ok(tx) = Transaction::deserialise(&s) {
                            txs.push(tx);
                        }
                    }
                }
                if !iter.advance() {
                    break;
                }
            }
        }
        txs
    }
}
